src/Interpreters/TransposedMetricLog.h

```diff
@@ -0,0 +1,74 @@
+#pragma once
+
+#include <Interpreters/PeriodicLog.h>
+#include <Common/ProfileEvents.h>
+#include <Common/CurrentMetrics.h>
+#include <Interpreters/SystemLog.h>
+#include <Core/NamesAndTypes.h>
+#include <Core/NamesAndAliases.h>
+#include <Storages/ColumnsDescription.h>
+
+#include <ctime>
+
+namespace DB
+{
+
+struct TransposedMetricLogElement
+{
+    UInt16 event_date;
+    time_t event_time;
+    Decimal64 event_time_microseconds{};
+    std::string metric_name;
+    Int64 value;
+    UInt8 is_event;
+
+    static std::string name() { return "TransposedMetricLog"; }
+    static ColumnsDescription getColumnsDescription();
+    static NamesAndAliases getNamesAndAliases() { return {}; }
+    void appendToBlock(MutableColumns & columns) const;
+};
+
+/// Transposed version of system.metric_log
+class TransposedMetricLog : public PeriodicLog<TransposedMetricLogElement>
+{
+    /// Optionally creates "wide" schema view for compatibility
+    std::string view_name;
+public:
+    static constexpr auto DESCRIPTION = R"(
+        Contains history of metrics values from tables system.metrics and system.events.
+        Periodically flushed to disk. Transposed form of system.metric_log.)";
+    static constexpr auto HOSTNAME_NAME = "hostname";
+    static constexpr auto EVENT_DATE_NAME = "event_date";
+    static constexpr auto EVENT_TIME_NAME = "event_time";
+    static constexpr auto EVENT_TIME_MICROSECONDS_NAME = "event_time_microseconds";
+    static constexpr auto METRIC_NAME = "metric";
+    static constexpr auto VALUE_NAME = "value";
+    static constexpr auto EVENT_TIME_HOUR_NAME = "toStartOfHour(event_time)";
+
+    static constexpr std::string_view PROFILE_EVENT_PREFIX = "ProfileEvent_";
+    static constexpr std::string_view CURRENT_METRIC_PREFIX = "CurrentMetric_";
+
+    /// This table is usually queried by time range + some fixed metric name.
+    static const char * getDefaultOrderBy() { return "event_date, toStartOfHour(event_time), metric"; }
+
+    void prepareTable() override;
+
+    /// We need to create view at startup, otherwise util first flush view will not exist
+    /// even if transposed table itself exists
+    bool mustBePreparedAtStartup() const override { return !view_name.empty(); }
```

Comment:
This is only if `schema_type = 'transposed_with_wide'` right?

Reply:
Yep, view_name is not empty only if it's `schema_type = 'transposed_with_wide'`. But it's the idea of this function.